### ISSUE TITLE
[ios] Initializing google map with 'n' markers takes O(n) time to add and blocks the UI during this period

### ISSUE THREAD

**Issue by bhavyalatha26** [flutter/flutter#60750]
## Details
Adding / initializing **'n'** markers on google map takes **O(n)** time, and the UI remains freezed during this period. I tested initializing with `10` , `100`, `500` and `1000` markers.
While adding `10` markers was pretty fast , the delay increased along with the number of markers with 1000 markers taking almost 5-15 seconds ! (in profile mode).  

The same code works fine on android , with no delay and is independent of size 'n'. The app reported no delay even in the debug mode. My android app is already live on Play store and no performance issue has been reported so far.

The issue is only with ios.

**Devices Tested :** 

- iPhone 5S (v 12.x)
- ios Simulator (v 13.x)

**Build Modes Tested :**

- debug
- profile

**Package**:
`google_maps_flutter 0.5.28+1`

**Code Snippet** :
```
Marker getMarkerForStation(Station s, {String infoText}) {
    return Marker(
        markerId: MarkerId(s.id),
        infoWindow: infoText == null ? InfoWindow.noText : InfoWindow(title: infoText),
        draggable: false,
        position: LatLng(s.geoLocation.latitude, s.geoLocation.longitude),
        onTap: () => _onTapExistingMarker(s)
  }

Widget _map(BuildContext context, List<Station> quickViewStations) {
    var _stationMarkers = Set.of(showStations.map((s) => getMarkerForStation(s)));
    return GoogleMap(
        markers: _stationMarkers,
        mapType: MapType.normal,
        myLocationEnabled: true,
        myLocationButtonEnabled: false,
        onMapCreated: _onMapCreated,
        mapToolbarEnabled: false,
        initialCameraPosition: CameraPosition(
          target: _currentDeviceLocation,
          zoom: 16.0,
        ),
        onCameraMove: _onCameraMove,
  }
```

**Target Platform:** ios
**Target OS version/browser:** 12 and above
**Devices:** ios mobile device / simulator

## Logs

```
Analyzing cng-eco-connect-flutter...                                    
No issues found! (ran in 5.4s)
```

```
[✓] Flutter (Channel master, 1.20.0-3.0.pre.51, on Mac OS X 10.15.4 19E2269, locale en-IN)
    • Flutter version 1.20.0-3.0.pre.51 at /Users/bhavyalathabandaru/Documents/Flutter/flutter
    • Framework revision c68758fab1 (6 days ago), 2020-06-24 21:23:02 -0700
    • Engine revision 0c14126211
    • Dart version 2.9.0 (build 2.9.0-18.0.dev d8eb844e5d)

 
[!] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/bhavyalathabandaru/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    ✗ Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] Connected device (2 available)
    • Bhavya i5s                 • 18dd1363a73d63042f402aa6f63b812fc966a094 • ios • iOS 12.4.4
    • iPhone SE (2nd generation) • 6C4BF3A7-8EC1-4E94-9BCA-850A3F41D170     • ios • com.apple.CoreSimulator.SimRuntime.iOS-13-5 (simulator)

! Doctor found issues in 1 category.
```

**Comment by TahaTesser** [comment]
Hi @bhavyalatha26 
I can reproduce the issue using 500 markers

https://imgur.com/BL8Bt0e

<details> 
<summary>Code Sample</summary> 

``` 
import 'dart:async';
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Google Maps Demo',
      home: MapSample(),
    );
  }
}

class MapSample extends StatefulWidget {
  @override
  State<MapSample> createState() => MapSampleState();
}

class MapSampleState extends State<MapSample> {
  Completer<GoogleMapController> _controller = Completer();

  Map<MarkerId, Marker> markers =
      <MarkerId, Marker>{}; // CLASS MEMBER, MAP OF MARKS

  static final CameraPosition _kGooglePlex = CameraPosition(
    target: LatLng(37.42796133580664, -122.085749655962),
    zoom: 0,
  );

  final random = Random();

  static final CameraPosition _kLake = CameraPosition(
      bearing: 192.8334901395799,
      target: LatLng(37.43296265331129, -122.08832357078792),
      tilt: 59.440717697143555,
      zoom: 19.151926040649414);

  Marker getMarkerForStation(Map<String, double> location, {String infoText}) {
    return Marker(
      markerId: MarkerId('5'),
      infoWindow:
          infoText == null ? InfoWindow.noText : InfoWindow(title: infoText),
      draggable: false,
      position: LatLng(location['lat'], location['lon']),
      // onTap: () => _onTapExistingMarker(s),
    );
  }

  @override
  void initState() {
    // TODO: implement initState
    super.initState();

    List<Map<String, double>> locations = [
      for (var i = 0; i < 500; i++)
        {
          "id": i.toDouble(),
          "lat": -90 + random.nextDouble() * 90 * 2,
          "lon": -180 + random.nextDouble() * 180 * 2
        },
    ];

    for (var item in locations) {
      final MarkerId markerId = MarkerId(item['id'].toString());
      // print(item);

      markers[markerId] = getMarkerForStation(item, infoText: 'text');
    }
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      body: GoogleMap(
        markers: Set<Marker>.of(markers.values),
        mapType: MapType.hybrid,
        initialCameraPosition: _kGooglePlex,
        onMapCreated: (GoogleMapController controller) {
          _controller.complete(controller);
        },
      ),
      floatingActionButton: FloatingActionButton.extended(
        onPressed: _goToTheLake,
        label: Text('To the lake!'),
        icon: Icon(Icons.directions_boat),
      ),
    );
  }

  Future<void> _goToTheLake() async {
    final GoogleMapController controller = await _controller.future;
    controller.animateCamera(CameraUpdate.newCameraPosition(_kLake));
  }
}

``` 

</details> 



<details> 
<summary>stable flutter doctor -v</summary> 

``` 
[✓] Flutter (Channel dev, 1.20.0-2.0.pre, on Mac OS X 10.15.5 19F101, locale
    en-GB)
    • Flutter version 1.20.0-2.0.pre at /Users/tahatesser/Code/flutter_dev
    • Framework revision 15a28159bc (10 days ago), 2020-06-23 04:52:58 -0700
    • Engine revision 91a63d6a44
    • Dart version 2.9.0 (build 2.9.0-19.0.dev 7e72c9ae7e)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.46.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.1

 
[✓] Connected device (5 available)            
    • SM M305F      • 32003c30dc19668f          • android-arm64  • Android 10
      (API 29)
    • Taha’s iPhone • 00008020-001059882212002E • ios            • iOS 13.5.1
    • macOS desktop • macos                     • darwin-x64     • Mac OS X
      10.15.5 19F101
    • Web Server    • web-server                • web-javascript • Flutter Tools
    • Chrome        • chrome                    • web-javascript • Google Chrome
      83.0.4103.116

• No issues found!
``` 

</details>

**Comment by bhavyalatha26** [comment]
Thank you for the response @TahaTesser . When could we expect a fix for this ? 
Actually I am waiting on it and cannot release my app on to the app store because of this issue. 

Also it would be helpful if you can suggest some workout if possible.

Regards,
Bhavya.

**Comment by chinmaygarde** [comment]
cc @csells @cyanglaz

**Comment by kf6gpe** [comment]
@cyanglaz, is this something we think we can get to soon?

**Comment by kf6gpe** [comment]
Followed up with @cyanglaz, who is working on a number of P1s & P2s at this point and likely won't get to this soon.

**Comment by kf6gpe** [comment]
@cyanglaz Do you want to add this to our list of Q4 goals? I don't know what else you have on your plate, though.

**Comment by cyanglaz** [comment]
@kf6gpe We probably won't have bandwidth to work on this in Q4, I'll add this to my queue tho in case I have some time.

**Comment by cyanglaz** [comment]
Had some time looking at this,

@bhavyalatha26 I don't think the time complexity could be reduced to less than O(n) as every single marker needs to be visited when adding them.
The performance issue on iOS is largely because raster thread and platform thread are merged when there is a map view.
PlatformViews(google maps in this case) in flutter does reduce performance since the rasterizer is running on the platform thread. 

I tried to see if we can put adding markers in the background thread but GoogleMap SDK seem to prevent us from doing so:
```
Terminating app due to uncaught exception 'GMSThreadException', reason: 'The API method must be called from the main thread'
```

Were you able to update your code such that only visible markers are displayed?
Also, marker clustering might help with this, although it has **not** been officially supported in the google_maps_flutter plugin, this is in our TODO list. See https://github.com/flutter/flutter/issues/26863

**Comment by bhavyalatha26** [comment]
Hi @cyanglaz thank you for the response. 

My use case is that , upon opening the page, all markers needs to be displayed in the map view soon after the map loads. So can you please suggest any other approach to load them (apart from clustering).

Regards,
Bhavya.

**Comment by cyanglaz** [comment]
@bhavyalatha26 Apart from clustering. I am not sure what would be the best solution. Maybe you could reduce the initial viewport of the map view?

**Comment by jl-gogovapps** [comment]
I think I'm having a similar issue with complex polygons.  I've been able to work around it by simplifying the geometry.  The map is stuck on gray for about 10-15 seconds before rendering successfully.  This happens on IOS only 

Would it be beneficial to have a separate ticket for efficiency in polygons, or do they fall in the same problem set (just being different symptoms)?

**Comment by tajchert** [comment]
I think this should be mentioned in README of maps plugin, we spent days to analyze why flutter map is lagging on iOS that much (and it is really visible to the user). 

In our case solution is:

- We limit loading only markers visible to the user by ViewBounds
- We will implement markers clustering (when the plugin will support it, #26863 )
- We can consider heatmap (when the plugin will support it, ready PR - https://github.com/flutter/plugins/pull/2454 )
- We will do a sample implementation using other maps plugins and comparison performance (will update when we will have some results) 

This is a real bummer and makes both user and developer experience pretty bad.

**Comment by TahaTesser** [comment]
<details> 
<summary>code sample</summary> 

```dart
import 'dart:async';
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Google Maps Demo',
      home: MapSample(),
    );
  }
}

class MapSample extends StatefulWidget {
  @override
  State<MapSample> createState() => MapSampleState();
}

class MapSampleState extends State<MapSample> {
  Completer<GoogleMapController> _controller = Completer();

  Map<MarkerId, Marker> markers =
      <MarkerId, Marker>{}; // CLASS MEMBER, MAP OF MARKS

  static final CameraPosition _kGooglePlex = CameraPosition(
    target: LatLng(37.42796133580664, -122.085749655962),
    zoom: 0,
  );

  final random = Random();

  static final CameraPosition _kLake = CameraPosition(
      bearing: 192.8334901395799,
      target: LatLng(37.43296265331129, -122.08832357078792),
      tilt: 59.440717697143555,
      zoom: 19.151926040649414);

  Marker getMarkerForStation(Map<String, double> location, {String infoText}) {
    return Marker(
      markerId: MarkerId('5'),
      infoWindow:
          infoText == null ? InfoWindow.noText : InfoWindow(title: infoText),
      draggable: false,
      position: LatLng(location['lat'], location['lon']),
      // onTap: () => _onTapExistingMarker(s),
    );
  }

  @override
  void initState() {
    // TODO: implement initState
    super.initState();

    List<Map<String, double>> locations = [
      for (var i = 0; i < 1500; i++)
        {
          "id": i.toDouble(),
          "lat": -90 + random.nextDouble() * 90 * 2,
          "lon": -180 + random.nextDouble() * 180 * 2
        },
    ];

    for (var item in locations) {
      final MarkerId markerId = MarkerId(item['id'].toString());
      // print(item);

      markers[markerId] = getMarkerForStation(item, infoText: 'text');
    }
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      body: GoogleMap(
        markers: Set<Marker>.of(markers.values),
        mapType: MapType.hybrid,
        initialCameraPosition: _kGooglePlex,
        onMapCreated: (GoogleMapController controller) {
          _controller.complete(controller);
        },
      ),
      floatingActionButton: FloatingActionButton.extended(
        onPressed: _goToTheLake,
        label: Text('To the lake!'),
        icon: Icon(Icons.directions_boat),
      ),
    );
  }

  Future<void> _goToTheLake() async {
    final GoogleMapController controller = await _controller.future;
    controller.animateCamera(CameraUpdate.newCameraPosition(_kLake));
  }
}

``` 

</details>

<details> 
<summary>logs</summary> 

```console
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[        ] ((null)) was false: Reached the max number of texture atlases, can not allocate more.
[        ] ((null)) was false: Failed to get the icon for the given CGImageRef.
[        ] ((null)) was false: Failed to allocate texture space for marker
[+3562 ms] [VERBOSE-2:profiler_metrics_ios.mm(203)] Error retrieving thread information: (os/kern) invalid argument
```

</details>

```yaml
dependencies:
  flutter:
    sdk: flutter
  cupertino_icons: ^0.1.3
  google_maps_flutter: ^2.0.3
  google_maps_flutter_web: ^0.3.0
```

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 2.0.5, on macOS 11.2.3 20D91 darwin-x64, locale
    en-GB)
    • Flutter version 2.0.5 at /Users/tahatesser/Code/flutter_stable
    • Framework revision adc687823a (6 days ago), 2021-04-16 09:40:20 -0700
    • Engine revision b09f014e96
    • Dart version 2.12.3

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.55.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (3 available)
    • Taha’s iPhone (mobile) • 00008020-001059882212002E • ios            • iOS
      14.4.2
    • macOS (desktop)        • macos                     • darwin-x64     •
      macOS 11.2.3 20D91 darwin-x64
    • Chrome (web)           • chrome                    • web-javascript •
      Google Chrome 90.0.4430.85

• No issues found!
```

```bash
[✓] Flutter (Channel master, 2.2.0-11.0.pre.215, on macOS 11.2.3 20D91
    darwin-x64, locale en-GB)
    • Flutter version 2.2.0-11.0.pre.215 at
      /Users/tahatesser/Code/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5865c14d67 (7 hours ago), 2021-04-22 06:44:03 +0200
    • Engine revision 7e69744ede
    • Dart version 2.14.0 (build 2.14.0-18.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[!] IntelliJ IDEA Community Edition (version unknown)
    ✗ Invalid IntelliJ version number.

[✓] VS Code (version 1.55.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (3 available)
    • Taha’s iPhone (mobile) • 00008020-001059882212002E • ios            • iOS
      14.4.2
    • macOS (desktop)        • macos                     • darwin-x64     •
      macOS 11.2.3 20D91 darwin-x64
    • Chrome (web)           • chrome                    • web-javascript •
      Google Chrome 90.0.4430.85

! Doctor found issues in 1 category.
```
	
</details>

**Comment by sperochon** [comment]
Hi TahaTesser,

For info, an old PR (11/2019) concerning the other older issue you just closed (#45508) was started here: https://github.com/flutter/plugins/pull/2302

**Comment by huycozy** [comment]
With regard to #123677 which is about **some markers that are not displayed on iOS** (in case there are a lot of markers), I also can reproduce this issue on the latest plugin version `google_maps_flutter: ^2.2.5` and Flutter master channel.

### Additional reports
- Android: the map can be dragged and zoomed better 
- iOS: 
    - dragging and zooming gestures are laggy
    - The log from output console is showing that it can not allocate more.
	```
	Debug service listening on ws://127.0.0.1:58843/CHbJinoP-Lo=/ws
	Syncing files to device iPhone...
	((null)) was false: Reached the max number of texture atlases, can not allocate more.
	((null)) was false: Failed to get the icon for the given CGImageRef.
	((null)) was false: Failed to allocate texture space for marker
	```

<details>
<summary>Demo</summary>

| iOS | Android |
| --------------- | --------------- |
<video src="https://user-images.githubusercontent.com/104349824/229020604-08197829-d64c-40b8-aa0a-d9c39176034d.MP4"/> | <video src="https://user-images.githubusercontent.com/104349824/229020620-7d1b76a4-de1e-4d1b-9077-f3af587fb740.mp4"/>
</details>

<details>
<summary>Sample code (1200 markers)</summary>

```dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
      debugShowCheckedModeBanner: false,
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final Completer<GoogleMapController> _controller =
  Completer<GoogleMapController>();

  @override
  void initState() {
    super.initState();
  }

  static const CameraPosition _cameraPosition = CameraPosition(
      target: LatLng(-35.28539099999, 149.13274999999), zoom: 5.0);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: GoogleMap(
        mapType: MapType.normal,
        rotateGesturesEnabled: true,
        zoomControlsEnabled: false,
        initialCameraPosition: _cameraPosition,
        onMapCreated: (GoogleMapController controller) {
          _controller.complete(controller);
        },
        markers: {
          //markers count
          for (var i = 0; i <= 1200; i++)
            Marker(
              markerId: MarkerId(i.toString()),
              position: LatLng(-35.28539099999 + i * 0.01, 149.13274999999 - i),
              draggable: true,
              onDragEnd: (value) {},
            ),
        },
      ),
    );
  }
}
```
</details>

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.7.8, on macOS 13.0.1 22A400 darwin-x64, locale en-VN)
    • Flutter version 3.7.8 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 90c64ed42b (34 hours ago), 2023-03-21 11:27:08 -0500
    • Engine revision 9aa7816315
    • Dart version 2.19.5
    • DevTools version 2.20.1

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14B47b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.76.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.60.0

[✓] Connected device (3 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos            • darwin-x64     • macOS 13.0.1 22A400 darwin-x64
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 111.0.5563.64

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```bash
[!] Flutter (Channel master, 3.9.0-18.0.pre.51, on macOS 13.0.1 22A400 darwin-x64, locale en-VN)
    • Flutter version 3.9.0-18.0.pre.51 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 659ba386f6 (3 hours ago), 2023-03-28 19:13:50 -0500
    • Engine revision 78f9c68971
    • Dart version 3.0.0 (build 3.0.0-376.0.dev)
    • DevTools version 2.22.2
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14B47b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.76.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.60.0

[✓] Connected device (3 available)
    • iPhone 8 Plus (mobile) • A6BF67EF-D048-430D-A097-324F146781DA • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)
    • macOS (desktop)        • macos                                • darwin-x64     • macOS 13.0.1 22A400 darwin-x64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 111.0.5563.110

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by rsachiendran** [comment]
Hi @huycozy, Is it possible to get an ETA when can this issue get solved? Cause this is a major feature used in our app and it is already in production, it is a critical issue in the app where it affects the user experience.

**Comment by huycozy** [comment]
>Is it possible to get an ETA when can this issue get solved?

@rsachiendran This issue has been prioritized with p4 which means the team agrees that it is important to work on. You might want to take a look at this: [How issue prioritization works](https://github.com/flutter/flutter/wiki/Issue-hygiene#priorities)

**Comment by rsachiendran** [comment]
Hi @huycozy, Is there a way to increase the priority?

**Comment by Colman** [comment]
> Had some time looking at this,
> 
> @bhavyalatha26 I don't think the time complexity could be reduced to less than O(n) as every single marker needs to be visited when adding them. The performance issue on iOS is largely because raster thread and platform thread are merged when there is a map view. PlatformViews(google maps in this case) in flutter does reduce performance since the rasterizer is running on the platform thread.
> 
> I tried to see if we can put adding markers in the background thread but GoogleMap SDK seem to prevent us from doing so:
> 
> ```
> Terminating app due to uncaught exception 'GMSThreadException', reason: 'The API method must be called from the main thread'
> ```
> 
> Were you able to update your code such that only visible markers are displayed? Also, marker clustering might help with this, although it has **not** been officially supported in the google_maps_flutter plugin, this is in our TODO list. See #26863

Do you know where the raster thread and platform thread are created? I've tried adding a timer to different parts of the code to figure out which call is taking the longest. I've found that setting the map property on the marker is what takes the most time. This call is found in the `GoogleMapMarkerController.m` file:
```obj-c
- (void)setVisible:(BOOL)visible {
  self.marker.map = visible ? self.mapView : nil;
}
```

When I time this call, I find that it takes about 4ms per marker. However, when I create a native iOS project with a Google Map and I call `marker.map = ...` it takes only 0.04ms. What I don't understand is why the times would be so different. I understand why adding the markers on the main thread would block the UI, but why would it change the amount of time the `marker.map` setter takes to execute? Isn't it the same computation in both cases? In the first case, is the Flutter engine switching to a different task part way through the execution of the `marker.map = ...` call which slows it down?

**Comment by Colman** [comment]
After some more investigation, I've found that there are two parts to this problem:

1. There is the slowness of rendering many markers
2. The rendering blocking the UI

Problem 1 is caused by this line in the `GoogleMapController`:
```dart
_mapView.accessibilityElementsHidden = NO;
```
This overrides the default behaviour of the GMSMapView which is `YES`. When I changed it to `YES`, the time to render 4000 markers went from 20 seconds to 2 seconds!

Problem 2 is caused by the main thread (Platform Task Runner) adding the markers to the map as discussed earlier. I believe the solution to this is to dispatch calls to the markers controller (addMarkers, changeMarkers, and removeMarkers) in a background task. Then just ensure that the calls are not made at the same time. This will prevent one call from modifying the markers list while it's being iterated by another.

I'm working on a PR to fix both of these issues now.

**Comment by OpenSphereSoftware** [comment]
Somehow i get this issue especially with dynamically generated Markers. It seems that BitmapDescriptor.bytes( somehow internally has a bug in it. BitmapDescriptor.asset( can do a lot more. 
@Colman really looking forward to your update :)

**Comment by Colman** [comment]
I've created the PR but it's still in draft since the my changes broke the UI tests. I turned off accessibility elements for performance reasons, but I think it needs to be enabled for UI testing. Does anyone know a workaround to this?

Some other things I've noticed are:

1. Using a background thread didn't really impact user experience after the other improvements. The UI blockage wasn't noticeable when adding/removing 2000 markers at once.

2. There's a very big difference between debug and release builds. When building in debug mode, I noticed the map got progressively slower as I added and removed markers. I would add 2000, remove them all, add 2000, remove them all, etc. The map would get much slower over time even though I had the same amount of markers. I reproduced this on the main branch as well as in an empty objective c project with just a Google Map. When I built in release mode, this didn't happen.

3. Adding an icon cache helped a lot. If you have 2000 markers that use only a few different icons, there's no sense in creating and sizing the icon for every marker. They should just be reused.

**Threading**
In general, setting any marker property (position, visibility, icon) has to be done on the main thread. If you are setting the property for the first time, such as when you're adding the marker, it will let you do it on the background thread. However, when you try to change the property from the background thread, like in changeMarkers, it gives an error. When I did my performance testing, I made sure to only call addMarkers and removeMarkers. In doing so, I was able to compare using a background thread vs. not. I found that there wasn't much of a difference either way after the other improvements.

**Comment by jokerttu** [comment]
@Colman note that this issue is visible on all platforms when using same bitmap multiple times.

> 3. Adding an icon cache helped a lot. If you have 2000 markers that use only a few different icons, there's no sense in creating and sizing the icon for every marker. They should just be reused.

As each time asset or bitmap is re-used on marker, it is processed and rescaled again. For this cache definitely would improve the performance.
For `BytesMapBitmap`, same bytes are sent to native layer to be processed, in some cases thousands of times. There is no real reason to send same PNG over the communication channel each time marker is added to map.

One possible solution would be to introduce new API:

Used by new `GoogleMapsFlutterPlatform` methods:
`Future<List<RegistryMapBitmap>>GoogleMapsFlutterPlatform.instance.registerMapBitmaps(List<BitmapDescriptor> bitmaps);`
`GoogleMapsFlutterPlatform.instance.unregisterMapBitmaps(List<RegistryMapBitmap> bitmaps);`
`GoogleMapsFlutterPlatform.instance.unregisterAllMapBitmaps();`

This would allow developers to process and cache bitmaps on before map is rendered, for example at app launch. Bitmaps can be then used also between each map view instances.

Communication with native layer can be done with new pigeon API: `GoogleMapsBitmapRegistryApi`. 
And have global bitmap registry instance on native implementations that handles the caching.
registerMapBitmaps should allow to register only certain types of BitmapDescriptor's: `bytes` and `asset`?

`RegistryMapBitmap` would extend `BitmapDescriptor`:
```dart
class RegistryMapBitmap extends BitmapDescriptor {
 RegistryMapBitmap._({
 required this.id,
 }) : super._(const <Object>[]);

 final RegistryMapId id;

 Static const String type = 'registry';

 @override
 Object toJson() => <Object>[
 type,
 <String, Object?>{
 'id': id,
 }
 ];
}
```

Not sure about need for following:
```
@immutable
class RegistryMapId extends MapsObjectId<RegistryMapBitmap> {
  /// Creates an immutable identifier for a [RegistryMapBitmap].
  const RegistryMapId(super.value);
}
```

Calling `registerBitmaps` would register/cache bitmaps on Android, iOS and web implementation for the API.

When marker has icon with `RegistryMapBitmap` type, it will fetch the image from cache with RegistryMapId.value to use the same, already preprocessed and scaled bitmap on native side. For this `BitmapDescriptor.registry(RegistryMapID id) `helper should be introduced.

This would allow developers to control caching, and cache invalidation; but as a downside the caching won't be available by default, if something like 
`static RegistryMapBitmap registryBytes(...)` is not introduced to `BitmapDescriptor` class.

Developers would need to be informed that they need to unregister the RegistryMapBitmap's on dispose to avoid memory leaks.

Not sure if this is good design; but this would give developers more control of the cache registry.

Before implementing it might be good ide to ask feedback of the design choices from the code owner (as this is not a small improvement).
